docs/blog/2021-09-07-mocking-a-production-api.md

```diff
@@ -0,0 +1,300 @@
+designing our POC solution.
+
+### 1. The Data Model
+
+The Events API returns a JSON payload which is a collection of Events. An
+example Event can be seen below:
+
+```json
+{
+    "arguments": "Ipod Nano - 8GB",
+    // Refers to a certain event and its resources.
+    "body": null,
+    // A text field containing information about the event.,
+    "created_at": "2015-04-20T08:33:57-11:00",
+    // The date and time (ISO 8601 format) when the event was created.
+    "id": 164748010,
+    // The ID of the event.
+    "desciption": "Received a new order",
+    // A human readable description of the event.
+    "path": "/admin/orders/406514653/transactions/#1145",
+    // A relative URL to the resource the event is for, if applicable.
+    "message": "Received a new order",
+    // A human readable description of the event. Can contain some HTML formatting.
+    "subject_id": 406514653,
+    // The ID of the resource that generated the event.
+    "subject_type": "Order",
+    // The type of the resource that generated the event.
+    "verb": "confirmed"
+    // The type of event that occurred. 
+}
+```
+
+Off the bat it's clear that there is some business logic that needs to be
+implemented. For example, there is some notion of causality, i.e. an `Order`
+cannot be `closed` before its been `placed`. This non-trivial business logic was
+good news - it means we can piggy back off some of the complex data generation
+logic that's built into Synth.
+
+Since we don't have access to the code that Runs the Shopify API, we have to
+simulate the behaviour of the data model. There are varying degrees of depth
+into which you can go, and we broke it into 4 levels:
+
+1. Level 1 - **Stub**: Level 1 is just about exposing and endpoint where the
+   data on a per element basis 'looks' right. You have the correct types, but
+   you don't really care about correctness across elements. For example you care
+   that `path` has the correct `subject_id` in the URI, but you don't care that
+   a given Order goes from `placed` to `closed` to `re_opened`etc...
+2. Level 2 - **Mock**: Level 2 involves maintaining the semantics of the Events
+   collection. For example `created_at` should be monotonically increasing as it
+   follows `id`. `verb`s should follow proper causality (as per the order
+   example above). etc.
+3. Level 3 - **Emulate**: Level 3 is about maintaining semantics across
+   endpoints. For example creating an order in a different Shopify API endpoint
+   should create an `order_placed` event in the Events API.
+4. Level 4 - **Simulate**: Here you are basically reverse engineering all the
+   business logic of the API. It should be indistinguisable from the real thing.
+
+Really these levels can be seen as increasing in scope as you go simulate
+semantics per-element, per-endpoint, cross-endpoint and finally for the entire
+API.
+
+### 2. The Endpoint Behaviour
+
+The Events API is not a naive CRUD system. The endpoint exposes various query
```

Comment:
I thought it was. The REST endpoints only has getters with, as you say after, filters.